tempesta_fw/http.c

```diff
@@ -742,6 +743,53 @@ tfw_http_error_resp_switch(TfwHttpReq *req, int status, const char *reason)
 	tfw_http_send_resp(req, code);
 }
 
+static inline void
+tfw_http_srv_hmonitor(TfwHttpResp *resp)
+{
+	u32 crc32;
+	char *body_cstr;
+	TfwServer *srv = (TfwServer *)resp->conn->peer;
+
+	if (!test_bit(TFW_SRV_B_HMONITOR, (unsigned long *)&srv->hm_flags))
+		return;
+
+	if (tfw_apm_hm_srv_limit(resp->status, srv->apmref)) {
+		if (!tfw_srv_suspended(srv)) {
+			tfw_srv_mark_suspended(srv);
+			TFW_WARN_ADDR("server has been suspended: limit"
+				      " for bad responses is exceeded",
+				      &srv->addr);
+		}
+		return;
+	}
+
+	if (!tfw_srv_suspended(srv))
+		return;
+
+	if (!tfw_apm_hm_srv_status_alive(resp->status, srv->apmref)) {
+		body_cstr = tfw_pool_alloc(resp->pool, resp->body.len + 1);
+		if (unlikely(!body_cstr)) {
+			TFW_ERR("Can't allocate memory for hmonitor"
+				" crc32 verification\n");
+			return;
+		}
+		tfw_str_to_cstr(&resp->body, body_cstr, resp->body.len + 1);
+		crc32 = crc32(0, body_cstr, resp->body.len);
```

Comment:
Frankly, this looks like extra unnecessary work. The whole body of a response is copied to memory allocated from pool just to calculate CRC32 on it.

This is a response to a health monitoring request, so I guess it's relatively small, and because it comes from a server it probably comes in minimal number of chunks (i.e. it's not sent one byte at a time). As far as I can see `crc32()` function is perfectly suited for calculating the checksum over data that is located in different chunks (the `seed` argument), so it should be straightforward to write a function that goes over string's chunks and calculates CRC32.

Reply:
Done.